dev/com.ibm.ws.microprofile.faulttolerance.2.0/src/com/ibm/ws/microprofile/faulttolerance20/impl/AsyncCompletionStageExecutor.java

```diff
@@ -63,16 +63,16 @@ protected void commitResult(AsyncExecutionContextImpl<CompletionStage<R>> execut
     }
 
     @Override
-    protected void finalizeAttempt(AsyncAttemptContextImpl<CompletionStage<R>> attemptContext, MethodResult<CompletionStage<R>> result) {
+    protected void processMethodResult(AsyncAttemptContextImpl<CompletionStage<R>> attemptContext, MethodResult<CompletionStage<R>> result, BulkheadReservation reservation) {
```

Comment:
Note switch from overriding `finalizeAttempt` to overriding `processMethodResult`.

`processMethodResult` occurs after the user's method has executed and has returned a result. When this happens, if the user's code has returned a CompletionStage, we want to defer processing until that CompletionStage is complete.

In contrast, `finalizeAttempt` is called when we've decided that the attempt is over for any reason (method complete, timeout, user cancellation, execution blocked by circuit breaker or bulkhead). Overriding this method was incorrect for several reasons:
1. The only situation where we want to defer processing is when the user's method completes
2. Releasing the bulkhead needs to occur when the method actually completes, not when the result is returned to the user (required by spec)